algorithms/kernel/svm/svm_train_thunder_impl.i

```diff
@@ -0,0 +1,338 @@
+#include "service/kernel/data_management/service_numeric_table.h"
+#include "service/kernel/service_utils.h"
+#include "service/kernel/service_data_utils.h"
+#include "externals/service_ittnotify.h"
+#include "externals/service_blas.h"
+#include "externals/service_math.h"
+
+#include "algorithms/kernel/svm/svm_train_common.h"
+#include "algorithms/kernel/svm/svm_train_thunder_workset.h"
+#include "algorithms/kernel/svm/svm_train_thunder_cache.h"
+#include "algorithms/kernel/svm/svm_train_result.h"
+
+#include "algorithms/kernel/svm/svm_train_common_impl.i"
+
+namespace daal
+{
+namespace algorithms
+{
+namespace svm
+{
+namespace training
+{
+namespace internal
+{
+template <typename algorithmFPType, typename ParameterType, CpuType cpu>
+services::Status SVMTrainImpl<thunder, algorithmFPType, ParameterType, cpu>::compute(const NumericTablePtr & xTable, NumericTable & yTable,
+                                                                                     daal::algorithms::Model * r, const ParameterType * svmPar)
+{
+    DAAL_ITTNOTIFY_SCOPED_TASK(COMPUTE);
+
+    services::Status status;
+
+    const algorithmFPType C(svmPar->C);
+    const algorithmFPType eps(svmPar->accuracyThreshold);
+    const algorithmFPType tau(svmPar->tau);
+    const size_t maxIterations(svmPar->maxIterations);
+    const size_t cacheSize(svmPar->cacheSize);
+    kernel_function::KernelIfacePtr kernel = svmPar->kernel->clone();
+
+    const size_t nVectors  = xTable->getNumberOfRows();
+    const size_t nFeatures = xTable->getNumberOfColumns();
+
+    TArray<algorithmFPType, cpu> alphaTArray(nVectors);
+    DAAL_CHECK_MALLOC(alphaTArray.get());
+    algorithmFPType * alpha = alphaTArray.get();
+
+    ReadColumns<algorithmFPType, cpu> mtY(yTable, 0, 0, nVectors);
+    DAAL_CHECK_BLOCK_STATUS(mtY);
+    const algorithmFPType * y = mtY.get();
+
+    // gradi = -yi; ai = 0
+    TArray<algorithmFPType, cpu> gradTArray(nVectors);
+    DAAL_CHECK_MALLOC(gradTArray.get());
+    algorithmFPType * grad = gradTArray.get();
+
+    for (size_t i = 0; i < nVectors; i++)
+    {
+        grad[i]  = -y[i];
+        alpha[i] = algorithmFPType(0);
+    }
+
+    TaskWorkingSet<algorithmFPType, cpu> workSet(nVectors, maxBlockSize);
+    DAAL_CHECK_STATUS(status, workSet.init());
+    const size_t nWS = workSet.getSize();
+    const size_t innerMaxIterations(nWS * cInnerIterations);
+
+    algorithmFPType diff     = algorithmFPType(0);
+    algorithmFPType diffPrev = algorithmFPType(0);
+
+    size_t innerIteration = 0;
+    size_t sameLocalDiff  = 0;
+
+    TArray<algorithmFPType, cpu> buffer(nWS * MemSmoId::latest + nWS * nWS);
+    DAAL_CHECK_MALLOC(buffer.get());
+
+    TArray<algorithmFPType, cpu> deltaAlpha(nWS);
+    DAAL_CHECK_MALLOC(deltaAlpha.get());
+
+    SVMCachePtr<thunder, algorithmFPType, cpu> cachePtr;
+
+    TArray<char, cpu> I(nWS);
+    DAAL_CHECK_MALLOC(I.get());
+
+    if (cacheSize >= nVectors * nVectors * sizeof(algorithmFPType))
```

Comment:
`nVectors * nVectors`  we have to check such computations on overflow according to SDL requirements.